sound/soc/sof/ipc4-topology.c

```diff
@@ -1485,6 +1485,59 @@ static int sof_ipc4_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget
 	return ret;
 }
 
+static int sof_ipc4_get_queue_id(struct snd_sof_widget *swidget, int bind_type,
+				 const char *peer_name)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_pin_binding *pin_binding;
+	struct ida *queue_ida;
+
+	int num_pins;
+	int i;
+
+	if (swidget->num_pin_bindings > 0) {
+		for (i = 0; i < swidget->num_pin_bindings; i++) {
+			pin_binding = &swidget->pin_binding[i];
+
+			if (pin_binding->bind_type == bind_type &&
+			    !strcmp(pin_binding->wname, peer_name)) {
+				return pin_binding->queue_id;
+			}
+		}
+		dev_err(scomp->dev, "failed to find queue id from pin binding array for %s\n",
+			swidget->widget->name);
+		return -EINVAL;
+	}
+
+	if (bind_type == SOF_ROUTE_BIND_SOURCE) {
+		queue_ida = &swidget->source_queue_ida;
+		num_pins = swidget->num_source_pins;
+	} else {
+		queue_ida = &swidget->sink_queue_ida;
+		num_pins = swidget->num_sink_pins;
+	}
+
+	if (num_pins < 1) {
+		dev_err(scomp->dev, "invalid num_pins %d for ida allocation for %s\n",
+			num_pins, swidget->widget->name);
+		return -EINVAL;
+	}
+
+	return ida_alloc_max(queue_ida, num_pins, GFP_KERNEL);
+}
+
+static void sof_ipc4_put_queue_id(struct snd_sof_widget *swidget, int queue_id, int bind_type)
+{
+	if (swidget->num_pin_bindings > 0)
+		return;
+
+	if (bind_type == SOF_ROUTE_BIND_SOURCE) {
+		ida_free(&swidget->source_queue_ida, queue_id);
+	} else {
+		ida_free(&swidget->sink_queue_ida, queue_id);
+	}
```

Comment:
this seems wrong, the code is not symmetrical between get and put. You always call ida_free() in the put, but don't always call ida_alloc_max()  in the get.

Reply:
the check `if (swidget->num_pin_bindings > 0) return` means we are using queue_id from topology, and returned, `ida_free()` is not called under this condition, I will add comment.